mne/preprocessing/xdawn.py

```diff
@@ -1,266 +1,362 @@
+from ..epochs import _BaseEpochs, EpochsArray
+from ..io import _BaseRaw
+from ..io.pick import _pick_data_channels
 from ..utils import logger
-from ..cov import _regularized_covariance
-from ..channels.channels import ContainsMixin
+from ..externals.six import iteritems
+
+
+def _construct_signal_from_epochs(epochs, events, sfreq, tmin):
+    """Reconstruct pseudo continuous signal from epochs."""
+    n_epochs, n_channels, n_times = epochs.shape
+    tmax = tmin + n_times / float(sfreq)
+    start = (np.min(events[:, 0]) + int(tmin * sfreq))
+    stop = (np.max(events[:, 0]) + int(tmax * sfreq) + 1)
+
+    n_samples = stop - start
+    n_epochs, n_channels, n_times = epochs.shape
+    events_pos = events[:, 0] - events[0, 0]
+
+    raw = np.zeros((n_channels, n_samples))
+    for idx in range(n_epochs):
+        onset = events_pos[idx]
+        offset = onset + n_times
+        raw[:, onset:offset] = epochs[idx]
+
+    return raw
 
 
-def _least_square_evoked(data, events, event_id, tmin, tmax, sfreq):
-    """Least square estimation of evoked response from data.
+def _least_square_evoked(epochs_data, events, tmin, sfreq):
+    """Least square estimation of evoked response from epochs data.
 
     Parameters
     ----------
-    data : ndarray, shape (n_channels, n_times)
-        The data to estimates evoked
-    events : ndarray, shape (n_events, 3)
+    epochs_data : array, shape (n_channels, n_times)
+        The epochs data to estimate evoked.
+    events : array, shape (n_events, 3)
         The events typically returned by the read_events function.
         If some events don't match the events of interest as specified
         by event_id, they will be ignored.
-    event_id : dict
-        The id of the events to consider
     tmin : float
         Start time before event.
-    tmax : float
-        End time after event.
     sfreq : float
         Sampling frequency.
 
     Returns
     -------
-    evokeds_data : dict of ndarray
-        A dict of evoked data for each event type in event_id.
-    toeplitz : dict of ndarray
-        A dict of toeplitz matrix for each event type in event_id.
+    evokeds_data : array, shape (n_class * n_components, n_channels)
+        An concatenated array of evoked data for each event type.
+    toeplitz : array, shape (n_class * n_components, n_channels)
+        An concatenated array of toeplitz matrix for each event type.
     """
-    nmin = int(tmin * sfreq)
-    nmax = int(tmax * sfreq)
-
-    window = nmax - nmin
-    n_samples = data.shape[1]
-    toeplitz_mat = dict()
-    full_toep = list()
-    for eid in event_id:
+
+    n_epochs, n_channels, n_times = epochs_data.shape
+    tmax = tmin + n_times / float(sfreq)
+
+    # Deal with shuffled epochs
+    events = events.copy()
+    events[:, 0] -= events[0, 0] + int(tmin * sfreq)
+
+    # Contruct raw signal
+    raw = _construct_signal_from_epochs(epochs_data, events, sfreq, tmin)
+
+    # Compute average evoked
```

Comment:
I'm not sure that's the correct term considering the overlap correction. Maybe more something like "Compute the independent evoked responses per condition, correcting for overlap".

(No evokeds are ever averaged, and no averaging is actually done.)